src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,620 @@
+            {
+                return true;
+            }
+
+            if (prefixLength1 == prefixLength2)
+            {
+                // If the lengths are the same, prefer the one with the higher match priority.
+                // But only if it's an item that would have been hard selected.  We don't want
+                // to aggressively select an item that was only going to be softly offered.
+
+                var item1Priority = CalculatePriority(item1);
+                var item2Priority = CalculatePriority(item2);
+
+                if (item1Priority > item2Priority)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private int CalculatePriority(EditorCompletion.CompletionItem item)
+        {
+            if (!item.Properties.TryGetProperty<CompletionItemSelectionBehavior>("SelectionBehavior", out var itemSelectionBehavior) ||
+                itemSelectionBehavior != CompletionItemSelectionBehavior.HardSelection)
+            {
+                return MatchPriority.Default;
+            }
+            else
+            {
+                return GetMatchPriority(item);
+            }
+        }
+
+        private EditorCompletion.FilteredCompletionModel HandleNormalFiltering(
+            ImmutableArray<EditorCompletion.CompletionItem> sortedList,
+            ITextSnapshot snapshot,
+            int? caretPosition,
+            string filterText,
+            ImmutableArray<EditorCompletion.CompletionFilterWithState> filters,
+            EditorCompletion.CompletionTriggerReason triggerReason,
+            List<FilterResult> itemsInList,
+            bool useSuggestionMode)
+        {
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                var listWithSelections = GetHighlightedList(itemsInList, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, filters, EditorCompletion.UpdateSelectionHint.NoChange, centerSelection: true, uniqueItem: default);
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+            if (completionService == null)
+            {
+                var listWithSelections = GetHighlightedList(itemsInList, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, filters, EditorCompletion.UpdateSelectionHint.NoChange, centerSelection: true, uniqueItem: default);
+            }
+
+            var matchingItems = itemsInList
+                .Where(r => r.MatchedFilterText)
+                .Select(t => GetOrCreateRoslynItem(t.CompletionItem))
+                .AsImmutable();
+
+            var chosenItems = completionService.FilterItems(snapshot.GetOpenDocumentInCurrentContextWithChanges(), matchingItems, filterText);
```

Comment:
Not "document" from above?

Reply:
@jason@jason-m.com , could you please re-phrase?

---
In reply to: [221394997](https://github.com/dotnet/roslyn/pull/29016#discussion_r221394997) [](ancestors = 221394997)